### ISSUE TITLE
Goaccess terminal dashboard shortcut keys not working.

### ISSUE THREAD

**Issue by jaggerwang** [allinurl/goaccess#185]
I'm connecting remote CentOS 6 server from iTerm 2 on Mac OS X. When I run goaccess command remotely, the dashboard display ok, and shortcut keys like 'h' and 'q' working correctly, but not the Expand and Scroll keys, I can only see the first screen content.

**Comment by ArondeParon** [comment]
Just wanted to let you know that I am experiencing this as well.

**Comment by allinurl** [comment]
@ArondeParon Interesting, could you run the following in your server and post the output after pressing the following keys:

    h
    j
    k
    o
    c
    s
    /
    <Ctrl> + f
    <Ctrl> + b
    g
    <Shift> + g
    n
    <Tab>
    <Shift> + <tab>
    <Arrow right>
    <Arrow down>
    <Arrow up>
    <Arrow left>
    <Enter>

Here's the c program, copy this into your server and name the file `keys.c`

    #include <curses.h>
    int main (void) {
      WINDOW *_window = initscr ();
      cbreak ();

      /* Accept all keys */
      keypad (_window, true);

      /* Don't echo things that are typed */
      noecho ();

      /* Don't display cursor */
      curs_set (0);

      for (;;) {
        printw ("Press a Key ");
        refresh ();
        int key = wgetch (_window);
        printw ("%c, %d \n", key, key);
      }
      endwin ();

      return 0;
    }

compile with:

    gcc keys.c -lncurses

**Comment by ArondeParon** [comment]
Hi,

Here is the output:

	Press a Key h, 104
	Press a Key j, 106
	Press a Key k, 107
	Press a Key o, 111
	Press a Key c, 99
	Press a Key s, 115
	Press a Key /, 47
	Press a Key ^F, 6
	Press a Key ^B, 2
	Press a Key g, 103
	Press a Key G, 71
	Press a Key n, 110
	Press a Key     , 9
	Press a Key a, 353
	Press a Key ^E, 261
	Press a Key ^B, 258
	Press a Key ^C, 259
	Press a Key ^D, 260
	Press a Key
	, 10

**Comment by allinurl** [comment]
I get exactly the same key codes. BTW, are you using zsh as well? Does this happen if you run goaccess locally in your mac?  

Also could  you please try again the following code, it should replicate goaccess' settings:

    #include <curses.h>
    int main (void) {
      initscr ();
      clear ();
      noecho ();
      halfdelay (10);
      nonl ();
      intrflush (stdscr, FALSE);
      keypad (stdscr, TRUE);
      curs_set (0);
      mousemask (BUTTON1_CLICKED, NULL);
      cbreak ();
      
      for (;;) {
        printw ("Press a Key ");
        refresh ();
        int key = wgetch (stdscr);
        
        printw ("%c, %d", key, key);
        switch (key) {
          case 'h': printw("` - h pressed`");  break;
          case 'g': printw("` - g pressed`");  break;
          case 'G': printw("` - G pressed`");  break;
          case 'n': printw("` - n pressed`");  break;
          case '/': printw("` - / pressed`");  break;
          case 106: printw("` - j pressed`");  break;
          case 107: printw("` - k pressed`");  break;
          case 115: printw("` - s pressed`");  break;
          case 2  : printw("` - Ctrl+b pressed`");  break;
          case 338: printw("` - pg+down pressed`");  break;
          case 339: printw("` - pg+up pressed`");  break;
          case 353: printw("` - shift+tab pressed`");  break;
          case 6  : printw("` - Ctrl+f pressed`");  break;
          case 99 : printw("` - c pressed`");  break;
          case 9: printw("` - tab pressed`");  break;
          case 0x0a:
          case 0x0d:
          case 32:
          case KEY_ENTER:
          printw("` - ENTER pressed`");
          break;
        } 
        printw ("\n", key, key);
      } 
      endwin ();
      
      return 0;
    } 

Thanks again.

**Comment by ArondeParon** [comment]
Hi,

Thanks for the assist! I'm using zsh on OSX and received the same output when compiling locally.
I compiled and ran your latest code on my CentOS machine and received the following output:

Press a Key h, 104` - h pressed`
    Press a Key j, 106` - j pressed`
    Press a Key k, 107` - k pressed`
    Press a Key o, 111
    Press a Key c, 99` - c pressed`
    Press a Key s, 115` - s pressed`
    Press a Key /, 47` - / pressed`
    Press a Key ^F, 6` - Ctrl+f pressed`
    Press a Key ^B, 2` - Ctrl+b pressed`
    Press a Key g, 103` - g pressed`
    Press a Key G, 71` - G pressed`
    Press a Key n, 110` - n pressed`
    Press a Key     , 9` - tab pressed`
    Press a Key a, 353` - shift+tab pressed`
    Press a Key ^E, 261
    Press a Key ^B, 258
    Press a Key ^C, 259
    Press a Key ^D, 260
    , 13` - ENTER pressed`

**Comment by aphorise** [comment]
This is most likely an issue with the user Locale and or terminal application preferences in use by user settings rather than a bug or issue with GoAccess.

I've also attempted to recreate this issue using GoAccess `0.9` within a bare minimal CentOS 5, OSX with a different language (Dutch) & `iTerm2`.
I did not adjust or set LANG or other environment variables related to language:
```shell
[user@cent5-zsh goaccess]$ locale
# LANG=
# LC_CTYPE="POSIX"
# LC_NUMERIC="POSIX"
# LC_TIME="POSIX"
# LC_COLLATE="POSIX"
# LC_MONETARY="POSIX"
# LC_MESSAGES="POSIX"
# LC_PAPER="POSIX"
# LC_NAME="POSIX"
# LC_ADDRESS="POSIX"
# LC_TELEPHONE="POSIX"
# LC_MEASUREMENT="POSIX"
# LC_IDENTIFICATION="POSIX"
# LC_ALL=
```
The system was configured with US defaults: 
```shell
[user@cent5-zsh goaccess]$ sudo yum install system-config-language && sudo system-config-language
#|--------------------------------------|
#| Select the language for the system.  │
#│           English (USA)              |
#│   | OK |                 | Cancel |  │
#|--------------------------------------|
```
My `iTerm2` profile was as follows:
![iTerm2 profile settings](http://i.imgur.com/l7yFmir.png "iTerm2 profile settings")

I then went through the following series:

| key sequence |
| --- |
| * __`1, enter`__ (expand top requests) |
|   - __`page-up, page-down`__ (section-scroll) |
|   - __`j, k`__ (section-scroll) |
| * __`1`__ (contract top requests) |
| * __`1, o`__ (expand top requests again) |
| * __`UP, DOWN`__ (view scroll keeping current section open). |
| * __`5, enter`__ (expand top visitors) |
|   - __`page-up, page-down`__ (section-scroll) |
|  - __`j, k`__ (section-scroll) |

I did not experience any problems so at this point the issue (if any) is not clear and neither is it detailed or elaborated well enough for delimitation.

I'd advise those involved to review their `locale` settings of their localhost / machine relative to those of the remote server and what maybe forward. So for example: 
```shell
[user@cent5-zsh goaccess]$ sudo grep 'LC_' /etc/ssh/sshd_config
# AcceptEnv LANG LANGUAGE LC_CTYPE LC_NUMERIC LC_TIME LC_COLLATE LC_MONETARY LC_MESSAGES
# AcceptEnv LC_PAPER LC_NAME LC_ADDRESS LC_TELEPHONE LC_MEASUREMENT
# AcceptEnv LC_IDENTIFICATION LC_ALL
#//-----------------------------
#// ^^^ I normally set forwarding / accepting of thse OFF!
```

**Comment by rauberdaniel** [comment]
I’m getting this issue in Terminal.app and GoAccess version 1.5 when I try to evaluate about 3000 files (AWS CloudFront logs).  The key commands work at first but after a short time (several seconds, not consistent) GoAccess seems to completely freeze. Maybe this is an issue with files not being closed after reading and therefore running into some OS limitation?

If I combine all these files into one single file (e.g. `cat *.log >> combined.log`) and load that with GoAccess afterwards, the issue does not seem to appear.

**Comment by allinurl** [comment]
@rauberdaniel Thanks for reporting this. Can I ask, are those 3000 logs being constantly updated or mostly static? Also, does it make any difference if you feed goaccess using tail as:

    tail -qF access.log access.1.log access.2.log | goaccess - --log-format=COMBINED

**Comment by rauberdaniel** [comment]
No, those files are static and not updated at all (therefore, they could be closed after reading).

I just tried feeding them via
```
cat *.log | goaccess - --log-format=CLOUDFRONT
```
and it seems to work fine without freezing.

What I also just realized is that simply using `goaccess *.log …` doesn't even load all the files, only about half the amount of requests are shown as compared to the command with `cat`. I think that could match to the possible issue of too many open files ( https://superuser.com/questions/433746/is-there-a-fix-for-the-too-many-open-files-in-system-error-on-os-x-10-7-1 ) as goaccess itself tries to keep reading those files and therefore doesn’t close them?

**Comment by allinurl** [comment]
Good question, goaccess doesn't keep the files opened.

I pushed a fix for this, it should work fine. Please feel free to build from [**development**](https://github.com/allinurl/goaccess#build-from-github-development) and let me know if that solves the issue on your end. You may still need to change your max number of files that can be opened in your OS. Ideally goaccess could resort to inotify, however it would need to work across all platforms, including BSDs, etc.
